1.  Create a function that takes an integer as an argument and returns "Even" for even numbers or "Odd" for odd numbers.

In [ ]:
def even_or_odd(number):
    if number % 2 == 0:
        return "Even"
    else:
        return "Odd"

number = int(input("Enter a number: "))
print(even_or_odd(number))



Enter a number: 5265478
Even


In [ ]:
def even_or_odd(number):
  return ["Even", "Odd"][number % 2]
number = int(input("Enter a number: "))
print(even_or_odd(number))

Enter a number: 125
Odd


In [ ]:
def even_or_odd(number):
    status = ""
    if number % 2 == 0:
        status = "Even"
    else:
        status = "Odd"

    return status

In [ ]:
even_or_odd = lambda number: "Odd" if number % 2 else "Even"


In [ ]:
even_or_odd = lambda number: ["Odd", "Even"][number % 2]

In [ ]:
def even_or_odd(number):
    even_or_odd = {0: "Even", 1: "Odd"}
    return even_or_odd[number % 2]

In [ ]:
def even_or_odd(number):
    if number % 2 == 0:
        return "Even"
    elif number/2 != int:
        return "Odd"

In [ ]:
def even_or_odd(number):
    if number/2 == int(number/2): return "Even"
    else: return "Odd"

2.  Write a function that takes in a string of one or more words, and returns the same string, but with all words that have five or more letters reversed (Just like the name of this Kata). Strings passed in will consist of only letters and spaces. Spaces will be included only when more than one word is present

In [ ]:
def spin_words(sentence):
    words = sentence.split()
    reversed_words = []
    for word in words:
        if len(word) >= 5:
            reversed_words.append(word[::-1])
        else:
            reversed_words.append(word) 5
    reversed_sentence = " ".join(reversed_words)

    return reversed_sentence

In [ ]:
def spin_words(sentence):
    # Your code goes here
    return " ".join([x[::-1] if len(x) >= 5 else x for x in sentence.split(" ")])

In [ ]:
def spin_words(sentence):
    words = [word for word in sentence.split(" ")]
    words = [word if len(word) < 5 else word[::-1] for word in words]
    return " ".join(words)

In [ ]:
def spin_words(sentence):
    return ' '.join(word if len(word)<5 else word[::-1] for word in sentence.split())

In [ ]:
def spin_words(sentence):
    return " ".join(x[::-1] if len(x) >= 5 else x for x in sentence.split())

In [ ]:
import re

def spin_words(sentence):
    # Your code goes here
    return re.sub(r"\w{5,}", lambda w: w.group(0)[::-1], sentence)

3. A Narcissistic Number (or Armstrong Number) is a positive number which is the sum of its own digits, each raised to the power of the number of digits in a given base. In this Kata, we will restrict ourselves to decimal (base 10).

For example, take 153 (3 digits), which is narcissistic:

    1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153
and 1652 (4 digits), which isn't:

    1^4 + 6^4 + 5^4 + 2^4 = 1 + 1296 + 625 + 16 = 1938
The Challenge:

Your code must return true or false (not 'true' and 'false') depending upon whether the given number is a Narcissistic number in base 10.

This may be True and False in your language, e.g. PHP.

Error checking for text strings or other invalid inputs is not required, only valid positive non-zero integers will be passed into the function.

In [ ]:
def narcissistic(number):
    digits = [int(digit) for digit in str(number)]
    num_digits = len(digits)
    narcissistic_sum = sum(digit ** num_digits for digit in digits)
    return narcissistic_sum == number


In [ ]:
def narcissistic(value):
    return value == sum(int(x) ** len(str(value)) for x in str(value))

In [ ]:
def narcissistic( value ):
    value = str(value)
    size = len(value)
    sum = 0
    for i in value:
        sum += int(i) ** size
    return sum == int(value)

In [ ]:
def narcissistic(value):
    return bool(value==sum([int(a) ** len(str(value)) for a in str(value)]))

4.Welcome.

In this kata you are required to, given a string, replace every letter with its position in the alphabet.

If anything in the text isn't a letter, ignore it and don't return it.

"a" = 1, "b" = 2, etc.

Example
Input = "The sunset sets at twelve o' clock."
Output = "20 8 5 19 21 14 19 5 20 19 5 20 19 1 20 20 23 5 12 22 5 15 3 12 15 3 11"

In [ ]:
def alphabet_position(text):
  result = []
  for char in text.lower():
    if "a" <= char <= "z":
      result.append(str(ord(char)  - ord("a") + 1))
  return " ".join(result)

In [ ]:
def alphabet_position(text):
    return ' '.join(str(ord(c) - 96) for c in text.lower() if c.isalpha())

In [ ]:
def alphabet_position(s):
  return " ".join(str(ord(c)-ord("a")+1) for c in s.lower() if c.isalpha())

5. Implement the function unique_in_order which takes as argument a sequence and returns a list of items without any elements with the same value next to each other and preserving the original order of elements.   
 unique_in_order('AAAABBBCCDAABBB') == ['A', 'B', 'C', 'D', 'A', 'B']
unique_in_order('ABBCcAD')         == ['A', 'B', 'C', 'c', 'A', 'D']
unique_in_order([1, 2, 2, 3, 3])   == [1, 2, 3]
unique_in_order((1, 2, 2, 3, 3))   == [1, 2, 3]

In [11]:
def unique_in_order(sequence):
    result = []
    previous = None
    for item in sequence:
        if item != previous:
            result.append(item)
            previous = item
    return result

result is an empty list where we store unique items.
previous keeps track of the last added item in result.
As we loop through sequence:
If item != previous, it's added to result.
Update previous to the current item for the next iteration.

In [ ]:
def unique_in_order(iterable):
    result = []
    prev = None
    for char in iterable[0:]:
        if char != prev:
            result.append(char)
            prev = char
    return result

6.Create a function that returns the sum of the two lowest positive numbers given an array of minimum 4 positive integers. No floats or non-positive integers will be passed.

For example, when an array is passed like [19, 5, 42, 2, 77], the output should be 7.

[10, 343445353, 3453445, 3453545353453] should return 3453455.

In [12]:
def sum_two_smallest_numbers(numbers):
    return sum(sorted(numbers)[:2])

In [ ]:
def sum_two_smallest_numbers(num_list):
    num_list.sort()
    return num_list[0] + num_list[1]

7. The rgb function is incomplete. Complete it so that passing in RGB decimal values will result in a hexadecimal representation being returned. Valid decimal values for RGB are 0 - 255. Any values that fall out of that range must be rounded to the closest valid value.

Note: Your answer should always be 6 characters long, the shorthand with 3 will not work here.

Examples (input --> output):
255, 255, 255 --> "FFFFFF"
255, 255, 300 --> "FFFFFF"
0, 0, 0       --> "000000"
148, 0, 211   --> "9400D3"

In [14]:
def rgb(r, g, b):
    def clamp(x):
        return max(0, min(x, 255))

    r = clamp(r)
    g = clamp(g)
    b = clamp(b)

    return f"{r:02X}{g:02X}{b:02X}"

In [ ]:
def rgb(r, g, b):
    round = lambda x: min(255, max(x, 0))
    return ("{:02X}" * 3).format(round(r), round(g), round(b))

In [ ]:
def limit(num):
    if num < 0:
        return 0
    if num > 255:
        return 255
    return num


def rgb(r, g, b):
    return "{:02X}{:02X}{:02X}".format(limit(r), limit(g), limit(b))

8. Complete the method/function so that it converts dash/underscore delimited words into camel casing. The first word within the output should be capitalized only if the original word was capitalized (known as Upper Camel Case, also often referred to as Pascal case). The next words should be always capitalized.

Examples
"the-stealth-warrior" gets converted to "theStealthWarrior"

"The_Stealth_Warrior" gets converted to "TheStealthWarrior"

"The_Stealth-Warrior" gets converted to "TheStealthWarrior"

In [15]:
def to_camel_case(text):
    if not text:
        return ""

    words = re.split(r"[-_]", text)
    return words[0] + ''.join(word.capitalize() for word in words[1:])

In [ ]:
def to_camel_case(text):
    return text[:1] + text.title()[1:].replace('_', '').replace('-', '')

In [ ]:
def to_camel_case(text):
    removed = text.replace('-', ' ').replace('_', ' ').split()
    if len(removed) == 0:
        return ''
    return removed[0]+ ''.join([x.capitalize() for x in removed[1:]])

9.The maximum sum subarray problem consists in finding the maximum sum of a contiguous subsequence in an array or list of integers:

max_sequence([-2, 1, -3, 4, -1, 2, 1, -5, 4])
should be 6: [4, -1, 2, 1]
Easy case is when the list is made up of only positive numbers and the maximum sum is the sum of the whole array. If the list is made up of only negative numbers, return 0 instead.

Empty list is considered to have zero greatest sum. Note that the empty list or array is also a valid sublist/subarray.

In [16]:
def max_sequence(arr):
    max_so_far = current_max = 0

    for x in arr:
        current_max = max(0, current_max + x)
        max_so_far = max(max_so_far, current_max)

    return max_so_far

10.Your task is to sort a given string. Each word in the string will contain a single number. This number is the position the word should have in the result.

Note: Numbers can be from 1 to 9. So 1 will be the first word (not 0).

If the input string is empty, return an empty string. The words in the input String will only contain valid consecutive numbers.

Examples
"is2 Thi1s T4est 3a"  -->  "Thi1s is2 3a T4est"
"4of Fo1r pe6ople g3ood th5e the2"  -->  "Fo1r the2 g3ood 4of th5e pe6ople"
""  -->  ""

In [ ]:
import re

def order(sentence):
    if not sentence:
        return ""

    words = sentence.split()
    # Sort words based on the digit found in each word
    sorted_words = sorted(words, key=lambda word: int(re.search(r'\d', word).group()))
    return " ".join(sorted_words)

In [ ]:
def order(words):
  return ' '.join(sorted(words.split(), key=lambda w:sorted(w)))